examples/dynamic_history_vehicles_replacement.py

```diff
@@ -0,0 +1,212 @@
+                vehicles_to_trap[veh_id] = (agent_id, agent_spec)
+                ctx["agents"][agent_id] = agent_spec.build_agent()
+
+            # Create missions for selected vehicles
+            veh_missions = scenario.create_dynamic_traffic_history_missions(
+                vehicles_to_trap.keys(), ctx["elapsed_sim_time"]
+            )
+
+            # Create traps for selected vehicles to be triggered immediately
+            smarts.trap_history_vehicles(vehicles_to_trap, veh_missions)
+
+        # Create a table of vehicle trajectory lengths, filtering out non-moving vehicles
+        traj_lens = []
+        for v_id in scenario.traffic_history.all_vehicle_ids():
+            traj = list(scenario.traffic_history.vehicle_trajectory(str(v_id)))
+            if [row for row in traj if row.speed != 0]:
+                traj_lens.append((v_id, len(traj)))
+
+        # Filter out trajectories with less than 100 timesteps
+        vehicle_candidates = [str(t[0]) for t in traj_lens if t[1] > 100]
+
+        assert len(vehicle_candidates) > 0
+
+        k = vehicles_to_replace_randomly
+        if k > len(vehicle_candidates):
+            logger.warning(
+                "vehicles_to_replace_randomly={} is greater than the number of vehicle candidates ({}).".format(
+                    vehicles_to_replace_randomly, len(vehicle_candidates)
+                )
+            )
+            k = len(vehicle_candidates)
+
+        # Initialize trigger and define initial context
+        context = {
+            "agents": {},
+            "elapsed_sim_time": 0.0,
+            "vehicle_candidates": vehicle_candidates,
+            "vehicles_to_replace_randomly": k,
+        }
+        trigger = Trigger(should_trigger, on_trigger, context=context)
+
+        dones = {}
+        observations = smarts.reset(scenario)
+        while not dones or not all(done for done in dones.values()):
+            # Step trigger with updated context
+            trigger.update(elapsed_sim_time=smarts.elapsed_sim_time)
+
+            # Get agents from current context
+            agents = trigger.context["agents"]
+
+            # Step simulation
+            actions = {
+                agent_id: agents[agent_id].act(agent_obs)
+                for agent_id, agent_obs in observations.items()
+            }
+            logger.debug(
+                "stepping @ sim_time={} for agents={}...".format(
+                    smarts.elapsed_sim_time, list(observations.keys())
+                )
+            )
+            observations, rewards, dones, infos = smarts.step(actions)
+
+            for agent_id in agents.keys():
+                if dones.get(agent_id, False):
+                    if not observations[agent_id].events.reached_goal:
+                        logger.warning(
+                            "agent_id={} exited @ sim_time={}".format(
+                                agent_id, smarts.elapsed_sim_time
+                            )
```

Comment:
I tend to prefer f-strings but using format is always fine as well as not not have to double check for py2.7.
```suggestion
                            f"agent_id={agent_id} exited @ sim_time={smarts.elapsed_sim_time}"
```